isthmus/src/main/java/io/substrait/isthmus/SubstraitRelVisitor.java

```diff
@@ -262,16 +266,69 @@ public Rel visit(org.apache.calcite.rel.core.Aggregate aggregate) {
     List<Grouping> groupings =
         sets.filter(s -> s != null).map(s -> fromGroupSet(s, input)).collect(Collectors.toList());
 
-    List<Measure> aggCalls =
+    // get GROUP_ID() function calls
+    List<AggregateCall> groupIdCalls =
+        aggregate.getAggCallList().stream()
+            .filter(c -> c.getAggregation().equals(SqlStdOperatorTable.GROUP_ID))
+            .collect(Collectors.toList());
+
+    List<AggregateCall> filteredAggCalls =
         aggregate.getAggCallList().stream()
+            // remove GROUP_ID() function calls
+            .filter(c -> !groupIdCalls.contains(c))
+            .collect(Collectors.toList());
+
+    List<Measure> aggCalls =
+        filteredAggCalls.stream()
             .map(c -> fromAggCall(aggregate.getInput(), input.getRecordType(), c))
             .collect(Collectors.toList());
 
-    return Aggregate.builder()
-        .input(input)
-        .addAllGroupings(groupings)
-        .addAllMeasures(aggCalls)
-        .build();
+    ImmutableAggregate.Builder builder =
+        Aggregate.builder().input(input).addAllGroupings(groupings).addAllMeasures(aggCalls);
+
+    if (groupings.size() > 1) {
+      // remove the grouping set index if there was no explicit GROUP_ID() function call
+      if (groupIdCalls.isEmpty()) {
+        int groupingExprSize =
+            groupings.stream()
+                .flatMap(g -> g.getExpressions().stream())
+                .distinct()
+                .collect(Collectors.toList())
+                .size();
+        builder.remap(Remap.offset(0, groupingExprSize + aggCalls.size()));
+      } else {
+        // remap grouping set index at the field positions where the GROUP_ID() function calls were
+        final int groupingFieldCount =
+            groupings.stream()
+                .flatMap(g -> g.getExpressions().stream())
+                .collect(Collectors.toList())
+                .size();
+        final int filterAggCallCount = aggCalls.size();
+        final Integer groupingSetIndex = groupingFieldCount + filterAggCallCount;
+
+        final List<Integer> groupingFieldIndices =
+            IntStream.range(0, groupingFieldCount).mapToObj(i -> i).collect(Collectors.toList());
+
+        final List<Integer> remap = new ArrayList<>(groupingFieldIndices);
```

Comment:
Only the `remap` ArrayList seems to be needed.

```suggestion
        final List<Integer> remap =
            IntStream.range(0, groupingFieldCount)
                .mapToObj(i -> i)
                .collect(Collectors.toCollection(ArrayList::new));
```